conn_pool_test.go

```diff
@@ -153,6 +154,57 @@ func TestPoolAcquireAndReleaseCycle(t *testing.T) {
 	releaseAllConnections(pool, allConnections)
 }
 
+func TestPoolNonBlockingConections(t *testing.T) {
+	t.Parallel()
+
+	maxConnections := 5
+	openTimeout := 1 * time.Second
+	dialer := net.Dialer{
+		Timeout: openTimeout,
+	}
+	config := pgx.ConnPoolConfig{
+		ConnConfig:     *defaultConnConfig,
+		MaxConnections: maxConnections,
+	}
+	config.ConnConfig.Dial = dialer.Dial
```

Comment:
Instead of actually dialing to microsoft, you might want to just write your own `Dial` function. Then you can do whatever you want inside. It will also allow you to test prepareStatements, resetCounts and such since you could precisely setup race conditions.